lib/std/debug.zig

```diff
@@ -3,6 +3,21 @@
 // This file is part of [zig](https://ziglang.org/), which is MIT licensed.
 // The MIT license requires this copyright notice to be included in all copies
 // and substantial portions of the software.
+
+//! std.debug contains functions for capturing and printing out stack traces as
+//! well as misc. functions such as `print`, `assert`, and `panicExtra`.
+//!
+//! Many parts of the implementations can be overriden via declarations in the
```

Comment:
So what is the right set of overrides?

I'm quite confident we want to be able to override `SymbolMap`, `attemptWriteLineFromSourceFile` (regardless of the unfortunate naming), and `captureStackTraceFrom`.

The `panicTerminate`, `getWriter`, and `getTTYConfig` function don't add much complexity while making this code somewhat more convenient. For instance, they allow for using `panicImpl` in the freestanding `single_threaded` context which avoids some code duplication and complexity on the part of the user. I can also see reasonable use cases for the functions on a supported OS. 

I don't see much need for the `formatStackTraceLine` function. I mostly added that because I thought it would make unit testing this code easier. But it could be useful more generally.

How much complexity do these overrides really add? Why do you think this is a mess? If the annoyance is with the block of lines picking which implementation to use, it should be possible to do that in a fully general way using metaprogramming. I just thought this would be nicer to read/use.

Reply:
> The panicTerminate, getWriter, and getTTYConfig function don't add much complexity while making this code somewhat more convenient. For instance, they allow for using panicImpl in the freestanding single_threaded context which avoids some code duplication and complexity on the part of the user. I can also see reasonable use cases for the functions on a supported OS.

This is the part that doesn't convince me.

My vision for the debug interface is much simpler:
* Define a `SymbolMap` interface (a-la `Allocator`)
* Turn the three main `SymbolMap` implementations into their own files (`debug/symbol_map_{unix,windows,darwin}.zig`)
* Have a single definition here:
```zig
pub const SymbolMapImpl = switch (target) {
    .linux, .freebsd, .openbsd => SymbolMapUnix,
    .windows => SymbolMapWindows,
    .macos => SymbolMapDarwin,
    .wasi => SymbolMapDummy, // or whatever wasi prefers
    .freestanding => ...; // pick the user-specified impl from root.config.SymbolMapImpl
    else => @compileError("ehh"),
}
```
* Define a `StacktraceCollect` interface
* Make `StackIterator` adhere to this interface (and move it to its own file)
* * Have a single definition here:
```zig
pub const StacktraceCollectImpl = switch (target) {
    .freestanding => ...; // pick the user-specified impl from root.config.StacktraceCollectImpl
    else => FrameWalkerCollectImpl, // the default frame-walking impl we use right now
}
```
* Write everything in terms of this two interfaces, the resulting code is generic and valid for every target
* Last but not least, don't provide too many knobs to make the default panic function usable everywhere as this imposes a huge burden on us (have to keep in mind every possible target or use-case when making modifications to it). The idea is to provide all the building blocks in `debug.zig` so that a freestanding user can write their own panic function in ten lines of code (or less).

I already have a branch where I applied this scheme with success, let me know if this would work for your use case.

Reply:
> * Define a `SymbolMap` interface (a-la `Allocator`)
> * Turn the three main `SymbolMap` implementations into their own files (`debug/symbol_map_{unix,windows,darwin}.zig`)
> * Have a single definition here:
> 
> ```zig
> pub const SymbolMapImpl = switch (target) {
>     .linux, .freebsd, .openbsd => SymbolMapUnix,
>     .windows => SymbolMapWindows,
>     .macos => SymbolMapDarwin,
>     .wasi => SymbolMapDummy, // or whatever wasi prefers
>     .freestanding => ...; // pick the user-specified impl from root.config.SymbolMapImpl
>     else => @compileError("ehh"),
> }
> ```
>

Yep, this part makes perfect sense to me (other than a few minor details). 

I don't see any reason to use a true interface with dispatching. Instead I think it makes for sense to just duck type it. This avoids some complexity while also allowing for easy usage of whatever error sets the user wants.

I'm vaguely leaning toward just defaulting to SymbolMapDummy. This means that freestanding will just print out addresses by default which seems sane to me. Though, perhaps this leads toward worse discoverability?


I didn't originally think of this because I didn't understand all of the layers used in `DebugInfo`, `openSelfDebugInfo`, `DebugModuleInfo` etc. After fixing and cleaning up that code. I now know of a nice way to "flatten" the hierarchy, purge some dead functions, and break things into separate files in order to accomplish this.

> * Define a `StacktraceCollect` interface
> * Make `StackIterator` adhere to this interface (and move it to its own file)
> * * Have a single definition here:
> 
> ```zig
> pub const StacktraceCollectImpl = switch (target) {
>     .freestanding => ...; // pick the user-specified impl from root.config.StacktraceCollectImpl
>     else => FrameWalkerCollectImpl, // the default frame-walking impl we use right now
> }
> ```

I don't see any reason to have an full interface. A single duck typed function seems just fine to me. But otherwise, yep. Also, the default implementation will work just fine on most freestanding platforms. Only the windows part is os specific, the rest is actually architecture and/or calling convention specific.

> * Write everything in terms of this two interfaces, the resulting code is generic and valid for every target

I think you also need `attemptWriteLineFromSourceFile`.

> * Last but not least, don't provide too many knobs to make the default panic function usable everywhere as this imposes a huge burden on us (have to keep in mind every possible target or use-case when making modifications to it).

Why do `panicTerminate`, `getWriter`, and `getTTYConfig` impose a huge maintenance burden? I have less (almost no) experience maintaining the standard library for a language, but the burden for these functions seems mild at worst.

>  The idea is to provide all the building blocks in `debug.zig` so that a freestanding user can write their own panic function in ten lines of code (or less).

There are a decent amount of subtleties in the existing `panicImpl` (particularly in the multithreaded case). For this reason, I think it should be preferred to stick with the default implementation whenever possible. I think coping and pasting `panicImpl` and just swapping out `panicTerminate`, `getWriter`, and (maybe) `getTTYConfig` will be common even outside of freestanding. But if you don't fully understand what is going on with `panicImpl`, it will be easy to cause bugs or poor output in the case of panic inside a panic or other edge cases. One approach would be to write a `panicImplGen` function which takes in various function and can be used to build a panic function. But generally functional programming isn't the style of choice on `zig`. I am not going to die on this hill, but I thought I would make the case for why I think these overrides are reasonable.


> I already have a branch where I applied this scheme with success, let me know if this would work for your use case.

Why didn't you say that when first saw this PR :/ ? Or did you just create the branch?
Regardless, at this point I think it makes more sense to proceed with what is currently here. I like to look at the branch regardless, so if you could link it that would be great.

Reply:
> > * Define a `SymbolMap` interface (a-la `Allocator`)
> > * Turn the three main `SymbolMap` implementations into their own files (`debug/symbol_map_{unix,windows,darwin}.zig`)
> > * Have a single definition here:
> > 
> > ```zig
> > pub const SymbolMapImpl = switch (target) {
> >     .linux, .freebsd, .openbsd => SymbolMapUnix,
> >     .windows => SymbolMapWindows,
> >     .macos => SymbolMapDarwin,
> >     .wasi => SymbolMapDummy, // or whatever wasi prefers
> >     .freestanding => ...; // pick the user-specified impl from root.config.SymbolMapImpl
> >     else => @compileError("ehh"),
> > }
> > ```
> 
> Yep, this part makes perfect sense to me (other than a few minor details).
> 
> I don't see any reason to use a true interface with dispatching. Instead I think it makes for sense to just duck type it. This avoids some complexity while also allowing for easy usage of whatever error sets the user wants.
> 

I don't see any reason not to define a good interface and use it.
Duck typing should *not* be your first choice in defining an API.

> I'm vaguely leaning toward just defaulting to SymbolMapDummy. This means that freestanding will just print out addresses by default which seems sane to me. Though, perhaps this leads toward worse discoverability?
> 
> I didn't originally think of this because I didn't understand all of the layers used in `DebugInfo`, `openSelfDebugInfo`, `DebugModuleInfo` etc. After fixing and cleaning up that code. I now know of a nice way to "flatten" the hierarchy, purge some dead functions, and break things into separate files in order to accomplish this.
> 
> > * Define a `StacktraceCollect` interface
> > * Make `StackIterator` adhere to this interface (and move it to its own file)
> > * * Have a single definition here:
> > 
> > ```zig
> > pub const StacktraceCollectImpl = switch (target) {
> >     .freestanding => ...; // pick the user-specified impl from root.config.StacktraceCollectImpl
> >     else => FrameWalkerCollectImpl, // the default frame-walking impl we use right now
> > }
> > ```
> 
> I don't see any reason to have an full interface. A single duck typed function seems just fine to me.

Ditto. You must really like ducks.

> But otherwise, yep. Also, the default implementation will work just fine on most freestanding platforms. Only the windows part is os specific, the rest is actually architecture and/or calling convention specific.

That was for the sake of showing how it would work. Note that in real world use-cases one would use more than a single strategy, eg. use a complex one if possible and fallback to stack walking.

> 
> > * Write everything in terms of this two interfaces, the resulting code is generic and valid for every target
> 
> I think you also need `attemptWriteLineFromSourceFile`.
> 

At this point I give up.
Printing stuff has nothing to do with collecting the stack frames and translating addresses to function names.
I don't care how you're presenting this info to the user, the default handler prints bits and pieces of the source file but there's no reason for other handlers to do so.
Eg. An embedded system won't have a copy of every source file onboard, we only care of the stack trace. Hell, I'll tell you more. You only need the addresses, the symbolication is usually performed offline using `addr2line` or similar tools.

> > * Last but not least, don't provide too many knobs to make the default panic function usable everywhere as this imposes a huge burden on us (have to keep in mind every possible target or use-case when making modifications to it).
> 
> Why do `panicTerminate`, `getWriter`, and `getTTYConfig` impose a huge maintenance burden? I have less (almost no) experience maintaining the standard library for a language, but the burden for these functions seems mild at worst.
> 

Because every time we add/remove something we'd have to think of how to let any hypothetical third-party user implement it, add yet another hook if needed, write down what the hook is supposed to do...

> > The idea is to provide all the building blocks in `debug.zig` so that a freestanding user can write their own panic function in ten lines of code (or less).
> 
> There are a decent amount of subtleties in the existing `panicImpl` (particularly in the multithreaded case). For this reason, I think it should be preferred to stick with the default implementation whenever possible. I think copying and pasting `panicImpl` and just swapping out `panicTerminate`, `getWriter`, and (maybe) `getTTYConfig` will be common even outside of freestanding. But if you don't fully understand what is going on with `panicImpl`, it will be easy to cause bugs or poor output in the case of panic inside a panic or other edge cases. One approach would be to write a `panicImplGen` function which takes in various function and can be used to build a panic function. But generally functional programming isn't the style of choice on `zig`. I am not going to die on this hill, but I thought I would make the case for why I think these overrides are reasonable.
> 

Give me a practical use case where this could be useful.
On most freestanding targets you don't need any multi threaded bullshit, if you really need it you can easily copy&paste the implementation and suit it to your needs, the code is well commented and it's definitely not rocket science. 

> > I already have a branch where I applied this scheme with success, let me know if this would work for your use case.
> 
> Why didn't you say that when you first saw this PR :/ ? Or did you just create the branch?
> Regardless, at this point I think it makes more sense to proceed with what is currently here. I would like to look at the branch regardless, so if you could link it that would be great.

I was interested in seeing an alternative implementation, the code has been sitting in my HDD for some months already but I couldn't be arsed to clean it up and submit it.

Reply:
> I don't see any reason not to define a good interface and use it.
> Duck typing should _not_ be your first choice in defining an API.

Sure, my first choice would be type classes/concepts/traits, but that's not how zig works.
How would you purpose handling different errors set types (aka the `Error` of  `io.Writer`)? Just use the global error set? Have a separate type parameter?

> > I don't see any reason to have an full interface. A single duck typed function seems just fine to me.
> 
> Ditto. You must really like ducks.

Again, how would you handle variable error sets from this function?

> At this point I give up.
> Printing stuff has nothing to do with collecting the stack frames and translating addresses to function names.

I'm trying my best here, but it seems to me like what you want out of this PR keeps shifting and is different than what I want out of this PR. I was actually mostly targeting being able to easily get full stack trace printing on freestanding (and general clean up). Right now its easy to print out a list of addresses, but trying to print function names, source lines, etc isn't so nice. I get that you could just use addr2line, but if this could be made easy I think it would be broadly useful (and I have seen interest from several other people in the zig community). 

> I don't care how you're presenting this info to the user, the default handler prints bits and pieces of the source file but there's no reason for other handlers to do so.
> Eg. An embedded system won't have a copy of every source file onboard, we only care of the stack trace. Hell, I'll tell you more. You only need the addresses, the symbolication is usually performed offline using `addr2line` or similar tools.

Check out:
https://github.com/andrewrk/clashos/blob/8ca226eb088d2a29f9a4875fd1245abb9842940b/src/debug.zig#L184-L193
That's the sort of use case I was targeting with this PR.

> Give me a practical use case where this could be useful.

You want debug output to also go to file in addition to stdout. Without override ability you would have to copy the whole panic implementation just to do this basic thing.

> On most freestanding targets you don't need any multi threaded bullshit, if you really need it you can easily copy&paste the implementation and suit it to your needs, the code is well commented and it's definitely not rocket science.

IMO, this is a bit of an anti-pattern which will lead to issues. But as I said before, I'm not trying to die on this hill. I'm just trying to make the case for why I think this is a good idea. I will take the overrides and move them to a separate PR. At this point, the PR has shifted substantially and these overrides (`getWriter`, `detectTTYConfig`, etc) are mostly unrelated to the other changes.

> I was interested in seeing an alternative implementation, the code has been sitting in my HDD for some months already but I couldn't be arsed to clean it up and submit it.

Fair enough.

Reply:
Overrides now live at https://github.com/rgreenblatt/zig/pull/1.

Reply:
> > I don't see any reason not to define a good interface and use it.
> > Duck typing should _not_ be your first choice in defining an API.
> 
> Sure, my first choice would be type classes/concepts/traits, but that's not how zig works.
> How would you purpose handling different errors set types (aka the `Error` of `io.Writer`)? Just use the global error set? Have a separate type parameter?

You can either define a standard error set in the interface that implementors have to respect or play a bit with the error set union operator. Check out `json.zig` to see how custom parsers having different and custom error sets are handled.

> 
> > > I don't see any reason to have an full interface. A single duck typed function seems just fine to me.
> > 
> > 
> > Ditto. You must really like ducks.
> 
> Again, how would you handle variable error sets from this function?
> 
> > At this point I give up.
> > Printing stuff has nothing to do with collecting the stack frames and translating addresses to function names.
> 
> I'm trying my best here, but it seems to me like what you want out of this PR keeps shifting and is different than what I want out of this PR. I was actually mostly targeting being able to easily get full stack trace printing on freestanding (and general clean up). Right now its easy to print out a list of addresses, but trying to print function names, source lines, etc isn't so nice. I get that you could just use addr2line, but if this could be made easy I think it would be broadly useful (and I have seen interest from several other people in the zig community).
> 

I've dealt with this debug code (and other similar code) for quite some time, I'm trying to design an API that suits my vision and fulfills your needs.
The current API is flexible enough to let a generic user implement a bare-bones panic handler printing a message, a list of addresses or a list of symbols. What I don't want is exposing our internal panic handler as part of the public API.

> > I don't care how you're presenting this info to the user, the default handler prints bits and pieces of the source file but there's no reason for other handlers to do so.
> > Eg. An embedded system won't have a copy of every source file onboard, we only care of the stack trace. Hell, I'll tell you more. You only need the addresses, the symbolication is usually performed offline using `addr2line` or similar tools.
> 
> Check out:
> https://github.com/andrewrk/clashos/blob/8ca226eb088d2a29f9a4875fd1245abb9842940b/src/debug.zig#L184-L193
> That's the sort of use case I was targeting with this PR.
> 

(That's not a great example as the way the DWARF infos are stitched is wrong and all the addressed are mangled, check the open PR for more info)
That's what I had in mind, ClashOS can define its own panic handler using the building blocks we provide.
Given a list of addresses it can find the `LineInfo` and print the source line using that function, with no need to let any of this pass by `debug` interface.

> > Give me a practical use case where this could be useful.
> 
> You want debug output to also go to file in addition to stdout. Without override ability you would have to copy the whole panic implementation just to do this basic thing.
> 

That's not something I'd expect to see in the wild and is largely orthogonal to the whole debug infrastructure refactoring.
Retargeting the default writer used here may be useful, I didn't object to that.

> > On most freestanding targets you don't need any multi threaded bullshit, if you really need it you can easily copy&paste the implementation and suit it to your needs, the code is well commented and it's definitely not rocket science.
> 
> IMO, copy pasting the panic implementation is a bit of an anti-pattern which will lead to issues. But as I said before, I'm not trying to die on this hill. I'm just trying to make the case for why I think this is a good idea. I will take the overrides and move them to a separate PR. At this point, the PR has shifted substantially and these overrides (`getWriter`, `detectTTYConfig`, etc) are mostly unrelated to the other changes.
> 
> > I was interested in seeing an alternative implementation, the code has been sitting in my HDD for some months already but I couldn't be arsed to clean it up and submit it.
> 
> Fair enough.

Reply:
> You can either define a standard error set in the interface that implementors have to respect or play a bit with the error set union operator. Check out `json.zig` to see how custom parsers having different and custom error sets are handled.

I will take a look.

> That's what I had in mind, ClashOS can define its own panic handler using the building blocks we provide.
> Given a list of addresses it can find the `LineInfo` and print the source line using that function, with no need to let any of this pass by `debug` interface.

So my plan was that user should be able to override the functions necessary to be able to format stack traces (fmtStackTrace in this PR). Am I correct in saying that you are thinking that something like ClashOS should instead instead reimplement `formatStackTraceLine` +  `StackTraceDumper.stackTrace`? If so, that use case doesn't allow for types in the standard library to use stack trace dumping. For instance, IMO `GeneralPurposeAllocator` should be able to use the same stack trace formatting (symbolizing, formating lines, etc) approch as `panic`

> Retargeting the default writer used here may be useful, I didn't object to that.

Huh, I thought you objected to exactly that. What set of overrides were you objecting to with the comment:

> Write everything in terms of this two interfaces, the resulting code is generic and valid for every target

Or to repeat my original question: what do you think the minimal set of overrides should be? My understanding was that you thought that minimal set of overrides should just be those 2 interfaces (which certainly doesn't include retargeting the writer).
Here is the full set of override potentially being considered: https://github.com/ziglang/zig/pull/8228#issuecomment-797823004. Note that the current state of this PR only has those two 2 interfaces (the linked comment in intentionally out of data right now).

Reply:
> So my plan was that user should be able to override the functions necessary to be able to format stack traces (fmtStackTrace in this PR). Am I correct in saying that you are thinking that something like ClashOS should instead instead reimplement `formatStackTraceLine` + `StackTraceDumper.stackTrace` (functions introduced in this PR)? If so, that use case doesn't allow for types in the standard library to use stack trace dumping and be supported in general. For instance, IMO `GeneralPurposeAllocator` should be able to use the same stack trace formatting (symbolizing, formating lines, etc) approch as `panic`. In addition to the two interfaces mentioned above, this requires one more interface: `attemptWriteLineFromSourceFile`.
> 

`StackTraceDumper` is a convenient wrapper to fetch a stack trace, symbolize it and print it somewhere.
I'm not a fan of this kind of wrappers but I see its practical value.

Now let me blow your mind, you don't need `attemptWriteLineFromSourceFile` to be part of the debug interface.
If you pass a `formatStackTraceLine` as a comptime parameter to `StackTraceDumper` the problem is gone, the lazy user can easily instance a `StackTraceDumper` with a function that formats the trace however they like, that prints the source by looking it up on the FS or using tricks like the one employed by ClashOS or, well, do nothing at all.

> Note that an implementation of `attemptWriteLineFromSourceFile` which always returns false is perfectly valid.
> 
> > Retargeting the default writer used here may be useful, I didn't object to that.
> 
> Huh, I thought you objected to exactly that. What set of overrides were you objecting to with the comment:
> 
> > Write everything in terms of this two interfaces, the resulting code is generic and valid for every target
> 

That was mostly about the use of `getStdErr` in `debug.print`, it might be useful for the user to override that.
On second thought the user may just get its own working writer and use `print` on that, not using `debug.print` it's no big deal and anyway it's something orthogonal to this PR.

> Or to repeat my original question: what do you think the minimal set of overrides should be? My understanding was that you thought that minimal set of overrides should just be those 2 interfaces (which certainly doesn't include retargeting the writer).

See above, you can have a retargetable `StackTraceDumper` and leave the default implementations alone without the many checks for overrides.
The two interfaces mentioned above do the heavy lifting, printing the stack is left as an exercise to the reader IMO. But if you want a `StackTraceDumper` for the lazy user let's add it into the mix.

> Here is the full set of overrides potentially being considered: [#8228 (comment)](https://github.com/ziglang/zig/pull/8228#issuecomment-797823004). Note that the current state of this PR only has those two 2 interfaces (the linked comment is intentionally out of data right now).

I'll have a pass over the new code later or tomorrow.

Reply:
> Now let me blow your mind, you don't need `attemptWriteLineFromSourceFile` to be part of the debug interface.
> If you pass a `formatStackTraceLine` as a comptime parameter to `StackTraceDumper` the problem is gone, the lazy user can easily instance a `StackTraceDumper` with a function that formats the trace however they like, that prints the source by looking it up on the FS or using tricks like the one employed by ClashOS or, well, do nothing at all.

Yeah, but that doesn't work for other types/libraries formating stack traces, this is what `GeneralPurposeAllocator` does.  You could instead pass `attemptWriteLineFromSourceFile` in the `GeneralPurposeAllocator` config, but this isn't a good idea. So I will add `attemptWriteLineFromSourceFile` back into the mix of overrides.

Also, note that StackTraceDumper isn't public. It's just a helper type `debug.zig`, see https://github.com/rgreenblatt/zig/blob/eef298510db95ccb9684e5d6bd716c6e7eec8b70/lib/std/debug.zig#L239-L241.
Users should use `fmtStackTrace` instead, see https://github.com/rgreenblatt/zig/blob/eef298510db95ccb9684e5d6bd716c6e7eec8b70/lib/std/debug.zig#L230-L237

> On second thought the user may just get its own working writer and use print on that, not using debug.print it's no big deal and anyway it's something orthogonal to this PR.
 
Yeah, for sure, I will keep that in a separate PR for consideration later.

Reply:
There's no middle ground, I'll send an alternate PR when I get some time.

Reply:
Oof. If that is a deal-breaker for you I can just remove it. (GPA just won't work for the general freestanding usecase).

Reply:
Also, if you do want to just send an alternate PR. I would recommend forking this one. The override stuff is now a pretty minor part of the PR, most of the diff is just general cleanup.